### ISSUE TITLE
RecursionError 

### ISSUE THREAD

**Issue by gwbischof** [intake/intake#412]
I get this error when trying to access a Catalog attribute that doesn't exist.

```In [9]: b.aaa                                                                                                                  
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
<ipython-input-9-94b8c43a1a22> in <module>
----> 1 b.aaa

~/anaconda3/envs/garrett2/lib/python3.7/site-packages/intake/catalog/entry.py in __getattr__(self, attr)
    119             return self.__dict__[attr]
    120         else:
--> 121             return getattr(self._get_default_source(), attr)
    122 
    123     def __dir__(self):

~/anaconda3/envs/garrett2/lib/python3.7/site-packages/intake/catalog/entry.py in _get_default_source(self)
     90     def _get_default_source(self):
     91         """Instantiate DataSource with default agruments"""
---> 92         if self._default_source is None:
     93             self._default_source = self()
     94         return self._default_source

... last 2 frames repeated, from the frame below ...

~/anaconda3/envs/garrett2/lib/python3.7/site-packages/intake/catalog/entry.py in __getattr__(self, attr)
    119             return self.__dict__[attr]
    120         else:
--> 121             return getattr(self._get_default_source(), attr)
    122 
    123     def __dir__(self):

RecursionError: maximum recursion depth exceeded
```

**Comment by martindurant** [comment]
Actually, I get
```
KeyError                                  Traceback (most recent call last)
~/code/intake/intake/catalog/base.py in __getattr__(self, item)
    330             try:
--> 331                 return self[item]  # triggers reload_on_change
    332             except KeyError:

~/code/intake/intake/catalog/base.py in __getitem__(self, key)
    401             return out
--> 402         raise KeyError(key)
    403

KeyError: 'blah'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-9-7c5a0c82e45c> in <module>
----> 1 intake.cat.blah

~/code/intake/intake/catalog/base.py in __getattr__(self, item)
    331                 return self[item]  # triggers reload_on_change
    332             except KeyError:
--> 333                 raise AttributeError(item)
    334         raise AttributeError(item)
    335

AttributeError: blah
```

**Comment by gwbischof** [comment]
I don't have a suggestion yet. I'll try to sort out why I'm getting the error and why it works well for you. Maybe its because I'm working with a catalog of catalogs.

**Comment by gwbischof** [comment]
with a catalog of catalogs:
catalog['cat1'].goodkey gives recursion error
catalog['cat1']().goodkey works correctly.

**Comment by martindurant** [comment]
Can you please write a snippet demonstrating this, since I am still seeing the expected AttributeError even when accessing nested catalogues.

**Comment by danielballan** [comment]
This turned out to be a downstream issue (in databroker) related to our experimental Entry <---> DataSource `__getattr__` hacking. Now resolved, and once we pin to `intake >= 0.6.0` the `__getattr__` hacking will no longer be necessary.

Test verifying that is resolved was merged in https://github.com/bluesky/databroker/pull/572
